### PR TITLE
network, netstat: Simplify status update

### PR DESCRIPTION
**Pull request by EdDev**
**What this PR does / why we need it**:

This change aims to simplify the status update logic for network
interface.

It follows the expected logic where:
- The interface status is populated with IP data from the configuration
cache (written to during the network setup).
- Afterwards, it gets updated by the data collected from the domain
configuration.
- Lastly, it gets updated by the (optional) guest-agent data.

With the new logic in place, the SR-IOV reporting is partially fixed to
not report interfaces when the guest-agent is missing.
A full fix should come in a following change that resolves
https://github.com/kubevirt/kubevirt/issues/7050

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
~~Depends on #6951 , only the last commit is to be reviewed.~~

Although the changes have been introduced gradually while developing, the end result seem too different from the original code. Therefore, it is recommended to go over both version of the code and see it indeed does the needed. The diff tooling will most likely not help.

Please also warn about missing coverage in unit tests if you notice any.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by oshoval** [comment]
Thanks

Since this area is delicate, 
and the e2e test of infoSource doesn't check IPs

please consider to manually see we don't break the mac change behavior by this change
the ips should not appear as they appeared here
https://github.com/kubevirt/kubevirt/pull/6744#issuecomment-968279660
(might worth to add assert on a follow PR in order to automate it)

additional checks i did manually were:
1. run without guest agent
2. stop guest agent after there were some reports
worth to consider trying them please as well (we can add automation later if needed)

File: pkg/network/setup/netstat.go [hunk 1]
```diff
@@ -67,106 +66,78 @@ func (c *NetStat) CachePodInterfaceVolatileData(vmi *v1.VirtualMachineInstance,
...
-					}
-				}
-			} else {
-				// If not present in vmi.Status.Interfaces, create a new one based on domain.Spec
-				newInterface = v1.VirtualMachineInstanceNetworkInterface{
-					MAC:  interfaceMAC,
-					Name: domainInterface.Alias.GetName(),
+		// Override status data with information from the guest-agent
+		for index := range vmi.Status.Interfaces {
+			vmiStatusInterface := &vmi.Status.Interfaces[index]
```
Comment (ormergi) [inline pkg/network/setup/netstat.go:120]: Is it possible to get the Status interface from `range` like on line 86?

File: pkg/network/setup/netstat.go [hunk 2]
```diff
@@ -67,106 +66,78 @@ func (c *NetStat) CachePodInterfaceVolatileData(vmi *v1.VirtualMachineInstance,
...
+			if guestAgentInterface, exists := guestAgentInterfacesByMac[vmiStatusInterface.MAC]; exists {
+				delete(guestAgentInterfacesByMac, vmiStatusInterface.MAC)
```
Comment (ormergi) [inline pkg/network/setup/netstat.go:122]: Why is it necessary to delete the item from `guestAgentInterfacesByMac` map?

**Review by ormergi — COMMENTED** [review]
Overall looks good! 
I have few question just for understanding (please see inline comments)

What you think about extraction the Status interfaces logic from domain to a function and another one from the guest-agent?
I think it will make the code more explanatory

File: pkg/network/setup/netstat.go [hunk 3]
```diff
@@ -67,106 +66,78 @@ func (c *NetStat) CachePodInterfaceVolatileData(vmi *v1.VirtualMachineInstance,
...
+					Name:       domainIfaceAlias,
+					MAC:        interfaceMAC,
+					InfoSource: netvmispec.InfoSourceDomain,
+				})
 			}
 		}
+	}
 
-		domainInterfaceStatusByMac := map[string]api.InterfaceStatus{}
+	if len(domain.Status.Interfaces) > 0 {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/setup/netstat.go:107]: Yes, as described in the method description: `domain.Status.Interfaces: interfaces reported by the guest agent (empty if Qemu agent not running).`

File: pkg/network/setup/netstat.go [hunk 4]
```diff
@@ -67,106 +66,78 @@ func (c *NetStat) CachePodInterfaceVolatileData(vmi *v1.VirtualMachineInstance,
...
+		guestAgentInterfacesByMac := map[string]api.InterfaceStatus{}
 		for _, domainInterfaceStatus := range domain.Status.Interfaces {
-			domainInterfaceStatusByMac[domainInterfaceStatus.Mac] = domainInterfaceStatus
+			guestAgentInterfacesByMac[domainInterfaceStatus.Mac] = domainInterfaceStatus
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/setup/netstat.go:110]: The short answer is: The best association between interfaces is the mac address, which should be unique.

The name is of the network, but that name is not always available. The guest-agent provides data from within the guest, where this network name does not exist. There is an association taking place between the interfaces reported by the guest-agent and the data from the domain configuration, but this does not cover all cases (e.g. SR-IOV).
Furthermore, that association should actually be dropped and the only processing should happen here.

Therefore, given no network name is provided by the guest-agent and we need to perform the matching of which iface belongs to which network, we need to use something that remains unique, like the mac address.

File: pkg/network/setup/netstat.go [hunk 5]
```diff
@@ -67,106 +66,78 @@ func (c *NetStat) CachePodInterfaceVolatileData(vmi *v1.VirtualMachineInstance,
...
 		}
 
-		existingInterfacesSpecByName := map[string]v1.Interface{}
+		vmiSpecInterfacesByName := map[string]v1.Interface{}
 		for _, existingInterfaceSpec := range vmi.Spec.Domain.Devices.Interfaces {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/setup/netstat.go:114]: Done

File: pkg/network/setup/netstat.go (hunk 1 above)
Comment (EdDev) [reply pkg/network/setup/netstat.go:120]: Yes, but we need `vmiStatusInterface` here anyway (need to use it multiple times to mutate the original item), so why bother?

File: pkg/network/setup/netstat.go (hunk 2 above)
Comment (EdDev) [reply pkg/network/setup/netstat.go:122]: On line 140, the interfaces that have no been processed here have special treatment.
I left this part as it original was because I did not have something better (or more accurately, it was not the problematic part that I tried to solve).

I guess I can invest some time and see how it can be eliminated.

**Review by EdDev — COMMENTED** [review]
> What you think about extraction the Status interfaces logic from domain to a function and another one from the guest-agent?
> I think it will make the code more explanatory

Yes, it can be improved for sure. After the base PR is merged I will rebase and see if it is an trivial thing or it merits its own PR. There are a lot of things to improve, my first priority here was to make the logic simpler.

**Comment by EdDev** [comment]
> Since this area is delicate, and the e2e test of infoSource doesn't check IPs

What does that mean exactly?

> please consider to manually see we don't break the mac change behavior by this change the ips should not appear as they appeared here [#6744 (comment)](https://github.com/kubevirt/kubevirt/pull/6744#issuecomment-968279660) (might worth to add assert on a follow PR in order to automate it)

How is it possible that this is not covered by unit and e2e tests?
If you have reviewed this change, do you see a problem?

> additional checks i did manually were:
>     1. run without guest agent
>     2. stop guest agent after there were some reports
> 
> worth to consider trying them manually please as well (we can add tier 2 automation later if needed, at least the 2nd which is bit longer)

Manual tests have no value, they may be correct now and incorrect a second later when some other PR gets merged.
I use them for debugging and understanding how things work.
So even if I do this, and say it to the world, I do not expect anyone, especially the reviewers to believe me.

I was under the impression we did include unit and e2e tests to cover what has been done in the code. If we missed something, it should be added sooner than later.
If the e2e test are too costly, then at least a unit test should be in place. Do you see such tests missing?

**Comment by oshoval** [comment]
> > Since this area is delicate, and the e2e test of infoSource doesn't check IPs
> 
> What does that mean exactly?

`Infosource VMI with 3 interfaces should have the expected entries in vmi status`
This test checks mac change
It normalize ips because the ips was not the focus of this change,
but since we saw that removing the merger from virt-launcher might break the expected behavior,
I just raised that it might worth to either see it isn't break by this change as well.
We can add on a follow PR an assert that the ips are expected to be empty on the `domain` interface after the mac is change
in order to protect (as mentioned above as well)

> 
> > please consider to manually see we don't break the mac change behavior by this change the ips should not appear as they appeared here [#6744 (comment)](https://github.com/kubevirt/kubevirt/pull/6744#issuecomment-968279660) (might worth to add assert on a follow PR in order to automate it)
> 
> How is it possible that this is not covered by unit and e2e tests? If you have reviewed this change, do you see a problem?

see please explanation above, the ips were not the main scope, we decided to normalize it, and it only later found it worth to check that the ips should be empty after mac change

I didn't review deeply yet, because saw this test fails and should be fixed first
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7051/pull-kubevirt-e2e-k8s-1.22-sig-network/1480438057151762432

> > additional checks i did manually were:
> > 1. run without guest agent
> > 2. stop guest agent after there were some reports
> > worth to consider trying them manually please as well (we can add tier 2 automation later if needed, at least the 2nd which is bit longer)
> 
> Manual tests have no value, they may be correct now and incorrect a second later when some other PR gets merged. I use them for debugging and understanding how things work. So even if I do this, and say it to the world, I do not expect anyone, especially the reviewers to believe me.
> 
> I was under the impression we did include unit and e2e tests to cover what has been done in the code. If we missed something, it should be added sooner than later. If the e2e test are too costly, then at least a unit test should be in place. Do you see such tests missing?

about the 1st
Since guest agent it enabled by default, i don't think we have a e2e test that disable it and run some tests (even if we add we need to think what we want to cover).
we do have some unit tests for no guest agent

about the 2nd we don't have e2e, i think that https://github.com/kubevirt/kubevirt/pull/6951 unit tests pretty covers the building blocks of this flow (by checking when there is guest agent report, and by checking there isn't)

Personally i usually run some corner cases ad hoc manually, some deserves adding e2e tests and then we usually add them,
but some don't imo.

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/763ce92ce14d0447aa03d281c54f89e647993a72..b971d43fd6cea0835e4124bfa8552b150a30a291): Fixed the fails in the e2e tests.

TODO: Cover this in the unit-tests.

**Comment by EdDev** [comment]
change: The simplification has been reworked, to make things hopefully simpler.

The PR description has been also updated to suggest the review strategy: Although the changes have been introduced gradually while developing, the end result seem too different from the original code. Therefore, it is recommended to go over both version of the code and see it indeed does the needed.
Please also warn about missing coverage in unit tests.

**Comment by oshoval** [comment]
Please consider splitting it to 2 commits, one that refactor to functions without logic change
and one that adds the new logic
It will be much easier and safer to review

**Comment by EdDev** [comment]
> Please consider splitting it to 2 commits, one that refactor to functions without logic change and one that adds the new logic It will be much easier and safer to review

If you refer to the partial fix, then this was not an explicit objective to fix it. It was an subsequent result of placing the flow in a simpler form. So I do not see how this can be split.

If you are referring to the overall flow that changed, which can be considered a logic change, I do not see how such a split is possible. The code reached the current state because everyone added one small piece to it, without adjusting the existing to support such additions in a logically simple manner. It had to be taken apart one step at a time and rebuilt.
Unlike in other refactoring, I saw no value in reflecting these steps, as it would have made it worse to the reviewer to try and track it. I suggest you follow my suggestion on how to review this PR, if you are interested in diving in.

**Comment by oshoval** [comment]
> > Please consider splitting it to 2 commits, one that refactor to functions without logic change and one that adds the new logic It will be much easier and safer to review
> 
> If you refer to the partial fix, then this was not an explicit objective to fix it. It was an subsequent result of placing the flow in a simpler form. So I do not see how this can be split.
> 
> If you are referring to the overall flow that changed, which can be considered a logic change, I do not see how such a split is possible. The code reached the current state because everyone added one small piece to it, without adjusting the existing to support such additions in a logically simple manner. It had to be taken apart one step at a time and rebuilt. Unlike in other refactoring, I saw no value in reflecting these steps, as it would have made it worse to the reviewer to try and track it. I suggest you follow my suggestion on how to review this PR, if you are interested in diving in.

I refer to only the last commit, i think some parts of it were stripped to functions (lets say just the ones that can be done easily without logic change), and can be a separate commit, 
and the rest can be on a commit on top of it

File: pkg/network/setup/netstat.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/network/setup/netstat.go:87]: or was presented but later disabled (since virt-launcher has caching of the guest-agent report)

we might want to consider changing this behavior in the future.
If guest agent is stopped, the cache will still report for ever the last known state

**Comment by EdDev** [comment]
change: Rebase

**Comment by oshoval** [comment]
> I tried locally #7053 + this PR and the test fails due to the change of 7053 that asserts there is no IP reported for the "original" interface, in the mac change scenario
> 
> we expect to see the first snippet, and not the 2nd one as appears in this explanation [#6744 (comment)](https://github.com/kubevirt/kubevirt/pull/6744#issuecomment-968279660)
> 
> please check it
> 
> I think we should add masquerade tests for the scenarios like infoSource test is doing before taking this PR, (for example mac change) in other to see there are no more regressions

Root cause is this
https://github.com/kubevirt/kubevirt/pull/7051/files#diff-c2b23c7cf6d14c6f6a9e12f1a63d6c8004ead588e9d4d6a369aeea15701ca086L180

at the past in case there were any of the conditions in the main loop, the interfaces from the pod were overriden
now they are append

File: pkg/network/setup/netstat.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netstat.go]: Thanks for explaining!
Maybe we should start to think of some kind of "primary-key" from Interfaces is order to simplify the mapping so we wont need to do intermediate associations, or worry about the case where the MAC is change.
If it make sense I can open an issue and we can discuss it there.

[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/setup/netstat.go]: I do not understand what you are trying to solve or how. If you want to suggest something, you can just send a PR.

**Comment by EdDev** [comment]
change: Rebase

**Review by oshoval — APPROVED** [review]
Thanks

File: pkg/network/vmispec/interface.go [hunk 6]
```diff
@@ -49,16 +49,41 @@ func IsPodNetworkWithMasqueradeBindingInterface(networks []v1.Network, ifaces []
 }
 
 func LookupInterfaceStatusByMac(interfaces []v1.VirtualMachineInstanceNetworkInterface, macAddress string) *v1.VirtualMachineInstanceNetworkInterface {
-	for _, iface := range interfaces {
-		if iface.MAC == macAddress {
-			iface := iface
-			return &iface
```
Comment (AlonaKaplan) [inline pkg/network/vmispec/interface.go:55]: I think the original intention was to avoid pointer to a slice item, it makes sense since the slice is not immutable.
What is the reason for the change?

**Review by oshoval — APPROVED** [review]
